cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot_dra_benchmark_test.go

```diff
@@ -0,0 +1,384 @@
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/framework"
+	featuretesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/features"
+
+	. "k8s.io/autoscaler/cluster-autoscaler/utils/test"
+)
+
+func createTestResourceSlice(nodeName string, devicesPerSlice int, slicesPerNode int, driver string, device resourceapi.BasicDevice) *resourceapi.ResourceSlice {
+	sliceId := uuid.New().String()
+	name := fmt.Sprintf("rs-%s", sliceId)
+	uid := types.UID(fmt.Sprintf("rs-%s-uid", sliceId))
+	devices := make([]resourceapi.Device, devicesPerSlice)
+	for deviceIndex := 0; deviceIndex < devicesPerSlice; deviceIndex++ {
+		deviceName := fmt.Sprintf("rs-dev-%s-%d", sliceId, deviceIndex)
+		deviceCopy := device
+		devices[deviceIndex] = resourceapi.Device{Name: deviceName, Basic: &deviceCopy}
+	}
+
+	return &resourceapi.ResourceSlice{
+		ObjectMeta: metav1.ObjectMeta{Name: name, UID: uid},
+		Spec: resourceapi.ResourceSliceSpec{
+			NodeName: nodeName,
+			Driver:   driver,
+			Pool: resourceapi.ResourcePool{
+				Name:               nodeName,
+				ResourceSliceCount: int64(slicesPerNode),
+			},
+			Devices: devices,
+		},
+	}
+}
+
+func createTestResourceClaim(requestsPerClaim int, devicesPerRequest int, driver string, deviceClass string) *resourceapi.ResourceClaim {
+	claimId := uuid.New().String()
+	name := fmt.Sprintf("claim-%s", claimId)
+	uid := types.UID(fmt.Sprintf("claim-%s-uid", claimId))
+	expression := fmt.Sprintf(`device.driver == "%s"`, driver)
+
+	requests := make([]resourceapi.DeviceRequest, requestsPerClaim)
+	for requestIndex := 0; requestIndex < requestsPerClaim; requestIndex++ {
+		requests[requestIndex] = resourceapi.DeviceRequest{
+			Name:            fmt.Sprintf("deviceRequest-%d", requestIndex),
+			DeviceClassName: deviceClass,
+			Selectors:       []resourceapi.DeviceSelector{{CEL: &resourceapi.CELDeviceSelector{Expression: expression}}},
+			AllocationMode:  resourceapi.DeviceAllocationModeExactCount,
+			Count:           int64(devicesPerRequest),
+		}
+	}
+
+	return &resourceapi.ResourceClaim{
+		ObjectMeta: metav1.ObjectMeta{Name: name, UID: uid, Namespace: "default"},
+		Spec: resourceapi.ResourceClaimSpec{
+			Devices: resourceapi.DeviceClaim{Requests: requests},
+		},
+	}
+}
+
+func allocateResourceSlicesForClaim(claim *resourceapi.ResourceClaim, nodeName string, slices ...*resourceapi.ResourceSlice) (*resourceapi.ResourceClaim, bool) {
+	allocatedDevices := make([]resourceapi.DeviceRequestAllocationResult, 0, len(claim.Spec.Devices.Requests))
+	sliceIndex, deviceIndex := 0, 0
+	requestSatisfied := true
+
+allocationLoop:
+	for _, request := range claim.Spec.Devices.Requests {
+		for devicesRequired := request.Count; devicesRequired > 0; devicesRequired-- {
+
+			for sliceIndex < len(slices) && deviceIndex >= len(slices[sliceIndex].Spec.Devices) {
+				sliceIndex++
+				deviceIndex = 0
+			}
+
+			if sliceIndex >= len(slices) {
```

Comment:
How could sliceIndex ever be greater than or equal to len(slices) if we only ever increment it (starting from zero) if sliceIndex is less than len(slices) (L97) ?

Or maybe I'm running into @towca's observation that this is a tricky nested boolean for loop inside two enumeration for loops inside a named for loop. :)